### DISCUSSION TITLE
Serialization and Checkpointing

### DISCUSSION THREAD

**Discussion by keceli** [NWChemEx/NWChemEx#52] [Ideas]
Serialization within NWChemEx is based on MADWorld and Cereal libraries. MADWorld archives support serializing distributed objects such as TiledArray (TA) tensors used in NWChemEx/LibChemist. Cereal provides support for JSON, XML, and portable binary archives. Currently, our checkpointing strategy is based on the `Cache` mechanism implemented in PluginPlay (PP). PP provide `Module` class and different computational chemistry methods within NWChemEx are implemented as compositions of multiple Modules. The input parameters and outputs of these Modules are stored in a Cache, which basically maps the input values (stored as a hash string) to the output values. While the `Cache` is mainly used for memoization to prevent running identical calculations and reduce the runtime, it can be used for checkpoint/restart since it stores the application state. The basic requirement for this strategy to work is that all module inputs should be hashable and all module outputs should be serializable. We implemented serialization and hashing functions for PluginPlay `Any` class that holds `Module` inputs and outputs in a type-safe manner. This strategy enables us to store the Cache in binary, JSON or XML formats as available through our serialization functions and portable to different architectures. However, I/O for writing or reading from the disk is not optimized in this approach. Our strategy for optimizations is to serialize large intermediate outputs such as amplitude tensors of coupled-cluster calculations in HDF5 format, which provides various optimizations for different storage hierarchies (such as node-local persistent memory or burst buffers) and file-systems. An alternative strategy we are exploring is to use the “Very Low Overhead Checkpointing System” (VeloC). VeloC provides multi-level checkpointing to use the faster but less reliable local storage to implement lighter resilience levels that hold the checkpoints. With multi-level checkpointing applications can survive some of the failures without using an external storage repository. However, VeloC currently does not support saving distributed objects such as TiledArray tensors. We provided VeloC developers with a working example of an iterative coupled-cluster doubles calculation that can checkpoint and restart by serializing/deserializing the TiledArray tensors in MADWorld binary format. Once VeloC supports checkpoint and restart for this small example, we will work on integrating VeloC for NWChemEx.

**Comment by ryanmrichard** [comment]
A couple of points: 

> We implemented serialization and hashing functions for PluginPlay Any class that holds Module inputs and outputs in a type-safe manner.

- This isn't Python where you can do introspection of an object. Being able to serializing/hash an `Any` instance is a necessary, but not sufficient condition for serialization/hashing. In fact the serialization/hashing of the `Any` is basically just a pass through. 
- While the `Any` holds them in a type-safe manner, that's a moot point because pretty much everything in C++ is type-safe. I think the property you are going for is "type-erased"

> Our strategy for optimizations is to serialize large intermediate outputs such as amplitude tensors of coupled-cluster calculations in HDF5

- IMHO this makes it seem like we just link to HDF5 and magically we get checkpointing. There's a lot of non-trivial details being left out. 

> An alternative strategy...

- I'm less familiar with VeloC than HDF5, but again I imagine we need to do some interfacing, which means there's more details needed.

I know this was in response to a DOE report, so maybe what you have is fine for the report, but I think it would be a useful exercise to start by listing what we need out of the I/O backends, and what the various I/O backends need from us. From there we can figure out what needs (if any) we have to implement ourselves and what sort of interfaces are going to be needed.
